src/server/modemwatcher.cpp

```diff
@@ -31,15 +33,37 @@ ModemWatcher::ModemWatcher(const QString objectPath, const QString interface, QO
     QObject(parent), m_objectPath(objectPath), m_interface(interface), m_interfaceAvailable(false)
 {
     QDBusConnection::systemBus().connect(OfonoConstants::OFONO_SERVICE, m_objectPath,
-                                         OfonoConstants::OFONO_MODEM_INTERFACE, "PropertyChanged",
-                                         this, SLOT(onModemPropertyChanged(QString, QDBusVariant)));
+                                         OfonoConstants::OFONO_MODEM_INTERFACE,
+                                         "PropertyChanged", this,
+                                         SLOT(onModemPropertyChanged(QString, QDBusVariant)));
 
+    m_ofonoWatcher = new QDBusServiceWatcher(OfonoConstants::OFONO_SERVICE,
+                                             QDBusConnection::systemBus(),
+                                             QDBusServiceWatcher::WatchForRegistration,
+                                             this);
+    connect(m_ofonoWatcher, SIGNAL(serviceRegistered(QString)),
+            this, SLOT(getProperties()));
+
+    if (QDBusConnection::systemBus().interface()->isServiceRegistered(OfonoConstants::OFONO_SERVICE))
```

Comment:
Yes, we are trying to fix a 25 second dbus timeout freeze when certain dbus services are not available. Three sync calls to the dbus daemon is not in the same order of magnitude :)

We have to do at least two synch queries when we check whether the dbus daemon manages 1) ofono 2) connman dbus APIs. The current code does two sync calls to the dbus daemon to check whether the ofono interface is up, we could combine those, but the complexity of the implementation would increase without significant improvement to functionality.

Currently caching the registration status of oFono and connman is not needed. In the case of oFono we simply ask for NITZ when timed starts if the interface is available, and then every time that the interface gets (re-)registered. Timed listens to oFono NITZ dbus signals, too.

connman handling is similar: we attempt to send a message to connman during startup and when automatic time update settings change, the message is sent every time that connman registers its dbus API (it does restart now and then).